app/test/testmerginapi.cpp

```diff
@@ -3116,3 +3116,195 @@ void TestMerginApi::testDownloadWithNetworkErrorRecovery()
+  QFile::remove( tempFilePath );
+
+  // 3 => empty file
+  tempFilePath = QDir::tempPath() + "/test_empty_config.json";
+  {
+    QFile file( tempFilePath );
+    QVERIFY( file.open( QIODevice::WriteOnly ) );
+    file.write( "" );
+    file.close();
+  }
+  config = MerginConfig::fromFile( tempFilePath );
+  QVERIFY( !config.isValid );
+  QCOMPARE( config.selectiveSyncEnabled, false );
+  QFile::remove( tempFilePath );
+
+  // 4 => file-not-found
+  tempFilePath = QDir::tempPath() + "/nonexistent_config.json";
+  config = MerginConfig::fromFile( tempFilePath );
+  QVERIFY( !config.isValid );
+  QCOMPARE( config.selectiveSyncEnabled, false );
+}
+
+void TestMerginApi::testHasLocalChangesWithSelectiveSyncEnabled()
+{
+  // temporary project directory
+  QTemporaryDir tempDir;
+  QVERIFY( tempDir.isValid() );
+  QString projectDir = tempDir.path();
+
+  // create a mergin‑config file enabling selective sync,
+  // so that files under "photos" are excluded from the sync
+  QString configPath = projectDir + "/mergin-config.json";
+  {
+    QFile configFile( configPath );
+    QVERIFY( configFile.open( QIODevice::WriteOnly ) );
+    configFile.write( "{\"input-selective-sync\": true, \"input-selective-sync-dir\": \"photos\"}" );
+    configFile.close();
+  }
+
+  // create two server files => one in "photos" folder (excluded), one in root folder (not excluded)
+  MerginFile serverExcluded;
+  serverExcluded.path = "photos/photo.jpg";
+  serverExcluded.checksum = "img_checksum";
+
+  MerginFile serverIncluded;
+  serverIncluded.path = "data.txt";
+  serverIncluded.checksum = "data_checksum";
+
+  QList<MerginFile> oldServerFiles;
+  oldServerFiles.append( serverExcluded );
+  oldServerFiles.append( serverIncluded );
+
+  // retrieve config file we wrote
+  MerginConfig config = MerginConfig::fromFile( configPath );
+  QVERIFY( config.isValid );
+  QCOMPARE( config.selectiveSyncEnabled, true );
+  QCOMPARE( config.selectiveSyncDir, QString( "photos" ) ); // verify selective sync folder
+
+  // first scenario => local files list exactly matches the non‑excluded server file
+  // the excluded file ("photos/photo.jpg") is ignored, and no local changes should be detected
+  QList<MerginFile> localFilesNoChange;
+  localFilesNoChange.append( serverIncluded );
+
+  bool result = mApi->hasLocalChanges( oldServerFiles, localFilesNoChange, projectDir, config );
+  QVERIFY( !result );
+
+  // second scenario => local file list contains a modified version of non‑excluded file
+  // the checksum of "data.txt" is different, and local changes should be detected
+  QList<MerginFile> localFilesChanged;
+  {
+    MerginFile modifiedIncluded = serverIncluded;
+    modifiedIncluded.checksum = "different_checksum";
+    localFilesChanged.append( modifiedIncluded );
+  }
+
+  result = mApi->hasLocalChanges( oldServerFiles, localFilesChanged, projectDir, config );
+  QVERIFY( result );
+}
+
+void TestMerginApi::testHasLocalProjectChanges()
```

Comment:
Interesting approach to generate project metadata here in test. Let's see if it will work fine in future, I think we can merge it now :)